src/platform/nxp/k32w/k32w0/ram_storage.c

```diff
@@ -0,0 +1,211 @@
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "FunctionLib.h"
+#include "ram_storage.h"
+
+#ifndef RAM_STORAGE_LOG
+#define RAM_STORAGE_LOG 0
+#endif
+
+#if RAM_STORAGE_LOG
+#include "fsl_debug_console.h"
+#define RAM_STORAGE_PRINTF(...)                                                                                                    \
+    PRINTF("[%s] ", __FUNCTION__);                                                                                                 \
+    PRINTF(__VA_ARGS__);                                                                                                           \
+    PRINTF("\n\r");
+#else
+#define RAM_STORAGE_PRINTF(...)
+#endif
+
+struct settingsBlock
+{
+    uint16_t key;
+    uint16_t length;
+} __attribute__((packed));
+
+static rsError ramStorageAdd(ramBufferDescriptor * pBuffer, uint16_t aKey, const uint8_t * aValue, uint16_t aValueLength)
+{
+    rsError error;
+    struct settingsBlock * currentBlock;
+    const uint16_t newBlockLength = sizeof(struct settingsBlock) + aValueLength;
+
+    if (*(pBuffer->ramBufferLen) + newBlockLength <= pBuffer->ramBufferMaxLen)
+    {
+        currentBlock         = (struct settingsBlock *) &pBuffer->pRamBuffer[*(pBuffer->ramBufferLen)];
+        currentBlock->key    = aKey;
+        currentBlock->length = aValueLength;
+
+        memcpy(&pBuffer->pRamBuffer[*(pBuffer->ramBufferLen) + sizeof(struct settingsBlock)], aValue, aValueLength);
+        *(pBuffer->ramBufferLen) += newBlockLength;
+
+        error = RS_ERROR_NONE;
+    }
+    else
+    {
+        error = RS_ERROR_NO_BUFS;
+    }
+
+    RAM_STORAGE_PRINTF("key = %d lengthWriten = %d err = %d", aKey, aValueLength, error);
+
+    return error;
+}
+
+rsError ramStorageGet(const ramBufferDescriptor * pBuffer, uint16_t aKey, int aIndex, uint8_t * aValue, uint16_t * aValueLength)
+{
+    uint16_t i           = 0;
+    uint16_t valueLength = 0;
+    uint16_t readLength;
+    int currentIndex = 0;
+    const struct settingsBlock * currentBlock;
+    rsError error = RS_ERROR_NOT_FOUND;
+
+    while (i < *(pBuffer->ramBufferLen))
+    {
+        currentBlock = (struct settingsBlock *) &pBuffer->pRamBuffer[i];
```

Comment:
This cast is scary.  settingsBlock is packed, so should not have alignment requirements.  However, if somebody removes __attribute__((packed)), this can definitely break on many platforms.

I would suggest we need to do something more explicit to enforce alignment.

Reply:
I understand your concern, I will try to find a better solution.